Source/WTF/wtf/glib/FileSystemGlib.cpp

```diff
@@ -185,30 +246,55 @@ PlatformFileHandle openFile(const String& path, FileOpenMode mode, FileAccessPer
         return invalidPlatformFileHandle;
 
     GRefPtr<GFile> file = adoptGRef(g_file_new_for_path(filename.data()));
-    GRefPtr<GFileIOStream> ioStream;
-    GFileCreateFlags permissionFlag = (permission == FileAccessPermission::All) ? G_FILE_CREATE_NONE : G_FILE_CREATE_PRIVATE;
-
-    if (failIfFileExists) {
-        ioStream = adoptGRef(g_file_create_readwrite(file.get(), permissionFlag, nullptr, nullptr));
-        return ioStream.leakRef();
+    GFileCreateFlags creationFlags = (permission == FileAccessPermission::All) ? G_FILE_CREATE_NONE : G_FILE_CREATE_PRIVATE;
+
+    if (mode == FileOpenMode::Read)
+        return G_SEEKABLE(g_file_read(file.get(), nullptr, nullptr));
+
+    // Embarrasingly, GIO doesn't have an atomic create-or-open function.
+    //
+    // The closest thing to create-or-open in GIO is g_file_replace() and g_file_replace_readwrite(), but they have some problems:
+    // (a) In the case of modifying an existing file, they return a new file in /tmp, doing a move swap when closing.
+    //     Even calling g_output_stream_flush() won't update the original file.
+    // (b) g_file_replace_readwrite() followed by a g_io_stream_close() will truncate the file, but we don't want ReadWrite to truncate.
+    //
+    // There is no pretty way out of this. Two naive solutions are:
+    // (a) We try to create, if it fails we try to open -> If the file is deleted between the two operations we fail.
+    // (b) We try to open, if it fails we try to create -> If the file is created between the two operations we fail.
+    //
+    // The most reliable thing we can do is spinlock retrying the other operation when we get a G_IO_ERROR_NOT_FOUND or G_IO_ERROR_EXISTS.
```

Comment:
Hm... this is pretty intense, but I can't think of anything better.

g_file_replace() is normally what you want because semantically you are atomically replacing the entire file with the contents of the stream when the stream is closed. So it's a true atomic replace: nothing gets written unless the whole thing gets written. But that indeed does not work very well if you just want to create the file atomically and then leave the stream open. :/